### PR TITLE
Avoid consuming messages before being able to process them

### PR DESCRIPTION
**Pull request by indrekj**
Previously we were taking all messages using Queue#subscribe and then
using a thread pool to process them. It is more useful to keep the
messages that cannot be processed yet in the queue. This allows adding
new consumers that can start processing them.

Prefetch value tells rabbitmq how many messages can consumers on a
channel be given at a time (before they have to acknowledge or reject
one of the earlier received messages).

### CODE REVIEW

**Comment by take-five** [comment]
Does it mean, that all messages will be processed sequentially? Don't you think it can reduce throughput?

My understanding of using `prefetch` is that it can limit concurrency level of message processing, but you still have to use thread pool or something similar. Otherwise you have to spawn X consumers to achieve X level of concurrency, and I'm not sure that it's more efficient than having a thread pool.

**Comment by willmore** [comment]
>This also allowed us to remove our own thread pools as now rabbitmq will take it care for us.

Which thread pools do you mean exactly?

**Comment by indrekj** [comment]
@take-five you are correct, thank you. Additional testing reveals it makes it sequential. Still need to have our own thread pool. 2am programming mistakes, oh well. Updating soon.

**Comment by indrekj** [comment]
Updated. Did a manual test as well. Created a responder that sleeps 1 second and has max_concurrency set to 2. Sent 6 messages. It processed 2 messages in parallel and took 3 seconds in total.

File: lib/freddy/adapters/bunny_adapter.rb [hunk 1]
```diff
@@ -46,10 +48,13 @@ def on_no_route(&block)
       end
 
       class Queue < Shared::Queue
-        def subscribe(&block)
-          @queue.subscribe do |info, properties, payload|
+        def subscribe(manual_ack: false, &block)
```
Comment (willmore) [inline lib/freddy/adapters/bunny_adapter.rb]: "manual" in manual_ack strikes me as an odd choice. What is "manual" about it?

File: lib/freddy/consumers/respond_to_consumer.rb [hunk 2]
```diff
@@ -1,8 +1,8 @@
 class Freddy
   module Consumers
     class RespondToConsumer
-      def initialize(consume_thread_pool, logger)
-        @consume_thread_pool = consume_thread_pool
+      def initialize(max_concurrency, logger)
```
Comment (willmore) [inline lib/freddy/consumers/respond_to_consumer.rb]: Why the change of signature? Keep `consume_thread_pool` as the arg and force the caller (`Freddy::initialize`) to set the `max_concurrency`. No reason to change things here.

File: lib/freddy.rb [hunk 3]
```diff
@@ -105,7 +105,8 @@ def respond_to(destination, &callback)
   #   end
   def tap_into(pattern, options = {}, &callback)
     @logger.debug "Tapping into messages that match #{pattern}"
-    @tap_into_consumer.consume(pattern, @connection.create_channel, options, &callback)
+    channel = @connection.create_channel(prefetch: @max_concurrency)
```
Comment (willmore) [inline lib/freddy.rb]: I find it confusing / misleading to use the variable `@max_concurrency` when specifying the prefetch.
I suggest creating a member variable `@prefetch_buffer_size` which is calculated in `initialize`.
Right now the calculation is trivial, but it might be adjust later. The point is, that code regarding prefetch buffer configuration outside of `initialize` will only refer to `@prefetch_buffer_size`.

File: lib/freddy/adapters/bunny_adapter.rb (hunk 1 above)
Comment (indrekj) [inline lib/freddy/adapters/bunny_adapter.rb]: This is 1-to-1 mapping with Bunny Queue#subscribe . http://rubybunny.info/articles/queues.html . Manual ack means that you manually let rabbitmq know when a message has been processed Otherwise it assumes that all messages are immediately processed.

File: spec/freddy/consumers/respond_to_consumer_spec.rb [hunk 4]
```diff
@@ -1,24 +1,74 @@
...
+
+      msg_count.times { deliver_message }
+
+      sleep default_sleep
+      expect(queue.message_count).to eq(msg_count - max_concurrency)
+
+      process_message
+      expect(queue.message_count).to eq(0)
+
+      # Release the final queued message before finishing the test to avoid
```
Comment (willmore) [inline spec/freddy/consumers/respond_to_consumer_spec.rb]: Can this be put in an `after` hook? It distracts from the rest of the test.

File: spec/freddy/consumers/respond_to_consumer_spec.rb [hunk 5]
```diff
@@ -1,24 +1,74 @@
...
+      end
+      default_sleep
+
+      expect(@message_received).to be_falsy
+    end
+  end
+
+  context 'when thread pool is full' do
+    let(:max_concurrency) { 1 }
+    let(:msg_count) { 2 }
```
Comment (willmore) [inline spec/freddy/consumers/respond_to_consumer_spec.rb]: Should `msg_count` be defined relative to max_concurrency?
```ruby
let(:msg_count) { max_concurrency + 1 }
```

**Comment by indrekj** [comment]
Updated

File: lib/freddy/adapters/bunny_adapter.rb (hunk 1 above)
Comment (willmore) [inline lib/freddy/adapters/bunny_adapter.rb]: Ok - it's still a peculiar choice of terminology. "Manual" conveys human intervention. 
In this case I would prefer `explicit_ack`, but as we are reusing underlying concepts, `manual_ack` will do.

File: lib/freddy.rb [hunk 6]
```diff
@@ -24,17 +24,21 @@ class Freddy
...
-  def initialize(connection, logger, consume_thread_pool)
+  def initialize(connection, logger, max_concurrency)
     @connection = connection
     @logger = logger
+    @prefetch_buffer_size = max_concurrency
 
-    @tap_into_consumer = Consumers::TapIntoConsumer.new(consume_thread_pool, @logger)
-    @respond_to_consumer = Consumers::RespondToConsumer.new(consume_thread_pool, @logger)
+    @tap_into_consumer = Consumers::TapIntoConsumer.new(
+      Thread.pool(@prefetch_buffer_size), @logger
```
Comment (take-five) [inline lib/freddy.rb]: Does it mean that we're gonna have 2 separate thread pools?

Comment (indrekj) [inline lib/freddy.rb]: https://github.com/salemove/freddy/pull/49/commits/38f678335f38528b5d0f6eddb0503506654c85c8 updated
I changed it so that each consumer has its own thread pool (usually an app has only one consumer anyway). I also documented the change. Sharing a thread pool between multiple consumers doesn't make sense because each consumer has its own channel (channels can't be shared because they are not thread safe in bunny) with prefetch enabled.

**Comment by willmore** [comment]
!squash
